### ISSUE TITLE
Problem with GCP Secret Manager and Spring Boot app

### ISSUE THREAD

**Issue by raszkiewicz** [GoogleCloudPlatform/spring-cloud-gcp#955]
Spring Boot (2.5.7) has problem to access password from the GCP Secret Manager using `spring-cloud-gcp-starter-secretmanager` ver 2.0.7 throwing error

```
AnnotationConfigApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'defaultFeignClientConfiguration': Injection of autowired dependencies failed; nested exception is org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [com.google.protobuf.ByteString$LiteralByteString] to type [java.lang.String]
```

for a passsword declared in the application.properties as

```
webservices.security.password=${sm://my-password}
```

when I will replace it with regular string or even env variable it will work fine.

Failing part of the code looks like:

```
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.cloud.client.loadbalancer.LoadBalancedRetryFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.retry.backoff.BackOffPolicy;
import org.springframework.retry.backoff.ExponentialBackOffPolicy;

import feign.Retryer;
import feign.auth.BasicAuthRequestInterceptor;
import feign.codec.ErrorDecoder;

/**
 * Default feign client configuration.  Includes retry policies, basic auth user name and password, and HTTP status decoder.
 * @author Greg Meyer
 * @since 6.0
 */
public class DefaultFeignClientConfiguration
{   
    @Value("${webservices.retry.backoff.multiplier:3}")
    protected double backoffMultiplier;
    
    @Value("${webservices.retry.backoff.initialBackoffInterval:100}")
    protected long initialBackoffInterval;  
    
    @Value("${webservices.retry.backoff.maxInterval:20000}")
    protected long maxInterval;     
    
    @Value("${webservices.security.basic.user.name:}")
    protected String user;  
    
    @Value("${webservices.security.basic.user.password:}")
    protected String pass;  
    
    /**
     * Creates an instance of the a the default HTTP status translator.
     * @return An instance of the a the default HTTP status translator
     */
    @Bean
    public ErrorDecoder feignClientErrorDecoder()
    {
        return new DefaultErrorDecoder();
    }
    
    /**
     * Creates an instance of BasicAuth interceptor configured with a username and password.   This bean is only created if the
     * "webservices.security.basic.user.name" property is set.
     * @return An instance of BasicAuth interceptor configured with a username and password
     */
    @Bean
    @ConditionalOnProperty(name="webservices.security.basic.user.name", matchIfMissing=false)
    public BasicAuthRequestInterceptor basicAuthRequestInterceptor() 
    {
        return new BasicAuthRequestInterceptor(user, pass);
    }   
    
    /**
     * Creates an instance of a back off policy used in conjuntion with the retry policy.
     * @return An instance of a back off policy
     */
    @Bean
    public LoadBalancedRetryFactory backOffPolciyFactory()
    {
        return new LoadBalancedRetryFactory() 
        {
            @Override
            public BackOffPolicy createBackOffPolicy(String service) 
            {
                final ExponentialBackOffPolicy backoffPolicy = new ExponentialBackOffPolicy();
                backoffPolicy.setMultiplier(backoffMultiplier);
                backoffPolicy.setInitialInterval(initialBackoffInterval);
                backoffPolicy.setMaxInterval(maxInterval);
                
                return backoffPolicy;
            }
        };      
    }
    
    /**
     * Creates a default http retry policy.
     * @return A default http retry policy.
     */ 
    @Bean
    public Retryer retryer() 
    {
        /*
         * Default retryer config
         */
        return new Retryer.Default(200, 1000, 5);
    }
}
```

Any thoughts?

**Comment by elefeint** [comment]
@raszkiewicz Did this work with Spring Boot 2.4.x in the past? Could you create a minimal sample demonstrating the problem?
This is most likely blocked by #149, since order in which the current Secret Manager configuration is applied is not very flexible.

Most likely, Feign autoconfiguration happens early on, before `GcpSecretManagerEnvironmentPostProcessor` had a chance to run and introduce `ByteString` converters.

**Comment by elefeint** [comment]
Registering custom converters at the correct phase will likely work. [This is what the converters should look like](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/v3.0.0/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/secretmanager/GcpSecretManagerEnvironmentPostProcessor.java#L36-L55).

**Comment by raszkiewicz** [comment]
Basically, the solution that works is to implement a Custom Converter which implements the Generic Conversion Service and register the Converter in the main method before calling `SpringApplication.run`. 

```
    public static void main(String[] args) 
    {
		((DefaultConversionService)DefaultConversionService.getSharedInstance()).addConverter(new CustomConverter());
		SpringApplication.run(STAApplication.class, args);
    }  
```

and custom converter:

```
@Component
public class CustomConverter implements GenericConverter {
    @Override
    public Set<ConvertiblePair> getConvertibleTypes() {
        return Collections.singleton(new ConvertiblePair(ByteString.class, String.class));
    }

    @Override
    public Object convert(Object source, TypeDescriptor sourceType,
                          TypeDescriptor targetType) {
        if (sourceType.getType() == String.class) {
            return source;
        }
        try {
            source = ((ByteString) source).toString("UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return source;
    }
}
```

**Comment by meltsufin** [comment]
Thanks for the update!

**Comment by RamonaMaria** [comment]
is this bug solved in the last release? what dependencies and versions should I use so that I don't have it anymore?
thanks

**Comment by elefeint** [comment]
@RamonaMaria A workaround was provided in [this comment](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/issues/955#issuecomment-1048101815).
Work on supporting `spring.config.import`, which will allow ordering of different configuration sources, is ongoing and expected to complete by next release.

**Comment by RamonaMaria** [comment]
Thanks for the update. I saw that solution and implemented it but unfortunately I still get that error:`AnnotationConfigApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'defaultFeignClientConfiguration': Injection of autowired dependencies failed; nested exception is org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [com.google.protobuf.ByteString$LiteralByteString] to type [java.lang.String]`

**Comment by elefeint** [comment]
@RamonaMaria Could you open a new issue, and provide a minimal reproducible issue?
